### ISSUE TITLE
Pyright failing to detect type compatibility with TypeVar bound to a Callable type

### ISSUE THREAD

**Issue by SandyChapman** [microsoft/pyright#3359]
I'm working with a set of existing type stubs that use a TypeVar with a bound Callable type to create Generic callables. I've reduced an issue I was seeing to a simple reproducible example:

```python
from typing import Any, Callable, Protocol, TypeVar

class SomeType:
    ...

CallableT = TypeVar("CallableT", bound=Callable[..., SomeType])

class MyCallable(Protocol[CallableT]):
    __call__: CallableT

class SomeGenericCallable(Protocol):
    def __call__(self, *args: Any, **kwargs: Any) -> SomeType: ...

def get_callable() -> MyCallable[SomeGenericCallable]: ...

x: Callable[..., SomeType] = get_callable()
```

In the above example, we define a TypeVar bound to Callable which should enforce any generic parameters to conform to the bound type. 

We make a generic protocol class, `MyCallable` using this TypeVar and set its `__call__` member's type to the generic type, which should result in `__call__` also conforming to the Callable bound type.

Next, we define another protocol that explicitly defines a `__call__` function that should be compatible with the `MyCallable` protocol as well.

Then, we define a method that returns a generic value.

The issue appears when we attempt to assign the result of the function back to a variable of the bound type.

![image](https://user-images.githubusercontent.com/3920249/164254854-39b44dd4-5170-4591-9d52-bab24c283b41.png)

```
main.py:16:30 - error: Expression of type "MyCallable[SomeGenericCallable]" cannot be assigned to declared type "(...) -> SomeType"
    Type "MyCallable[SomeGenericCallable]" cannot be assigned to type "(...) -> SomeType" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 information
```

If instead of having the `MyCallable`'s `__call__` member be of a generic type, but instead using the bound type directly, no error is emitted.

```python
class MyCallable(Protocol):
    __call__: Callable[..., SomeType]
```

I believe this is a bug where Pyright should be detecting that `MyCallable[SomeGenericCallable]` is compatible with `Callable[..., SomeType]`.

**Comment by SandyChapman** [comment]
@erictraut You're correct. I did not write this code. I've reduce some of the Django type stubs to this example. If you're interested, you can see the original pattern at the following links:

https://github.com/typeddjango/djangorestframework-stubs/blob/master/rest_framework-stubs/views.pyi#L38-L49
https://github.com/typeddjango/djangorestframework-stubs/blob/master/rest_framework-stubs/viewsets.pyi#L28-L31
https://github.com/typeddjango/django-stubs/blob/master/django-stubs/urls/conf.pyi#L11-L21

With the example code that generated the error being:

```python
path(
    "my/path/here,
    views.MyViewSetSubclass.as_view({"put": "update_or_create"}),
)
```

If I had a better handle on the Django stubs, I'd open a fix there, but I'm unsure of the wider implications of changing their stubs.

**Comment by erictraut** [comment]
Upon further investigation, I think pyright is correct here. The problem is that the `get_callable` function returns a callable object, but the `__call__` attribute is itself a callable object rather than an actual function. This is easier to see if we make use instantiable classes rather than protocols.

```python
from typing import Any, Callable, Generic, TypeVar

class SomeType: ...
CallableT = TypeVar("CallableT", bound=Callable[..., SomeType])

class MyCallable(Generic[CallableT]):
    def __init__(self, cb: CallableT):
        self.__call__ = cb

    __call__: CallableT

class SomeGenericCallable:
    def __call__(self, *args: Any, **kwargs: Any) -> SomeType:
        return SomeType()
```

If we run the following, the program crashes with the error `MyCallable` is not callable.
```python
def get_callable():
    return MyCallable(SomeGenericCallable())

x = get_callable()
x()
```

If we replace the above with the following, the program runs without crashing:
```python
def get_callable():
    return SomeGenericCallable()

x = get_callable()
x()
```

So, as you can see, a `__call__` attribute must be populated with a function, not with another object that has a `__call__` method.

**Comment by SandyChapman** [comment]
@erictraut Thanks for the reply. I'm not sure this is the full picture though. From what you've said (that `__call__` must be a function) makes sense for runtime execution to be successful, but the fact that the class is a Protocol means that it's not intended to be executed directly and implementers of the protocol should be ensuring conformance.

However, even if we accept that `__call__` must be a function, Pyright should then be issuing an error in this case:

```python
from typing import Callable, Protocol, TypeVar

class SomeType:
    ...

CallableT = TypeVar("CallableT", bound=Callable[..., SomeType])

class MyCallable1(Protocol[CallableT]):
    __call__: CallableT

class MyCallable2(Protocol):
    __call__: Callable[..., SomeType]

def get_callable1() -> MyCallable1: ...
def get_callable2() -> MyCallable2: ...

x1: Callable[..., SomeType] = get_callable1() # Error here
x2: Callable[..., SomeType] = get_callable2() # No error here
```

In this example, we've replaced the `TypeVar` bound to `Callable[..., SomeType]` with `Callable[..., SomeType]` directly in `MyCallable2`'s definition. Again, `__call__` is not a function in this case, but a member with a `Callable` type. Pyright _does not_ emit an error. This appears to be inconsistent behaviour.

**Comment by erictraut** [comment]
I've fixed the apparent inconsistency. The assignments in the last two lines of the code sample above type check without errors.

But I don't think the assignment to `x1` does what you are expecting. The function `get_callable1` returns a value of type `MyCallable1`. The class `MyCallable1` is generic and accepts one type argument. Since you haven't specified a type argument, it is effectively `MyCallable1[Any]`, which means that the `__call__` class variable has the type `Any`.

**Comment by SandyChapman** [comment]
@erictraut Thanks, again. And also for pointing out that in simplifying the example I may have introduced something unexpectedly. In the actual code I'm working with, the type isn't generic, but a subclass that inherits a generic base, so I think this still addresses my issue. I'll test it out in the next version and open a new ticket if I see anything wrong.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.240, which I just published. It will also be included in the next release of pylance.

**Comment by philipn** [comment]
I'm currently seeing the `djangorestframework-stubs` error referenced by @SandyChapman.  I'm running pyright 1.1.299 and the relevant portions are:

  * `as_view()`, defined here: https://github.com/typeddjango/djangorestframework-stubs/blob/80a81fd9cce1cd4d989340b35bb158ff07a360e5/rest_framework-stubs/views.pyi#L60
  * Then usage in a Django `urls.py` file:

```
from django.urls import path
from myapp import views

urlpatterns = [
    path('path/here', views.MyAPIView.as_view(), name='my-api-view'),
]
```

Pyright emits a type error:

```
error: Argument of type "AsView[GenericView]" cannot be assigned to parameter "view" of type "List[URLResolver | str]" in function "path"
"AsView[GenericView]" is incompatible with "List[URLResolver | str]" (reportGeneralTypeIssues)
```

Which means the relevant `@override` for `path()` is not being matched (see https://github.com/typeddjango/django-stubs/blob/master/django-stubs/urls/conf.pyi#L20):

```
@overload
def path(
    route: str, view: Callable[..., HttpResponseBase], kwargs: dict[str, Any] = ..., name: str = ...
) -> URLPattern: ...
```

Using the mock code:

```
from typing import TypeVar, Protocol, Mapping, Any

class HttpResponseBase:
    pass

class Response(HttpResponseBase):
    pass

class APIView:
    pass

class HttpRequest:
    pass

_View = TypeVar("_View", bound=Callable[..., HttpResponseBase])

class AsView(Protocol[_View]):
    cls: type[APIView]
    view_class: type[APIView]
    view_initkwargs: Mapping[str, Any]
    csrf_exempt: bool
    __call__: _View

# Call signature for view function that's returned by as_view()
class GenericView(Protocol):
    def __call__(self, request: HttpRequest, *args: Any, **kwargs: Any) -> Response: ...
```

When I ask if `isinstance(AsView[GenericView], Callable)`, the answer is `True`.  So I am not sure if pyright's behavior here - not matching - is correct or not.

For what it's worth, mypy does the match correctly and there are no type errors.

**Comment by erictraut** [comment]
It took me a while to understand your post above, but I think I pieced together your code samples to form a self-contained minimal repro.

```pythonfrom django.urls import path
from typing import Callable, TypeVar, Protocol, Mapping, Any

class HttpResponseBase: ...
class Response(HttpResponseBase): ...
class HttpRequest: ...

_View = TypeVar("_View", bound=Callable[..., HttpResponseBase])

class AsView(Protocol[_View]):
    __call__: _View

class GenericView(Protocol):
    def __call__(self, request: HttpRequest, *args: Any, **kwargs: Any) -> Response: ...

def func(v: AsView[GenericView]):
    urlpatterns = [
        path("path/here", v, name="my-api-view"),
    ]
```

This code is problematic for the same reason that I explained earlier in this thread. A `__call__` method is not allowed to be populated with an object that has a `__call__` method. The runtime will reject this. That means the `AsView` protocol definition will not work as it's defined. The `__call__` method should be defined with a `def` statement, not as as variable. I think pyright is doing the right thing here. It's consistent with the runtime behavior, whereas mypy is not.
